### PR TITLE
Some improvements to modular reduction routines

### PR DESCRIPTION
**Pull request by hanno-becker**
The NIST curves `secp{192,224,256}` use dedicated modular quasi-reduction routines which reduce a double-size output `A0,...,A(2n-1)` of a modular multiplication down to a single-size representative `A'0,...,A'(n-1)`. They do so through repeatedly applying the underlying prime's equation to express each `A'(i)` as a linear combination of `A(i)` and some `A(n + j)`.

At the end of such a reduction round there may be a signed carry-out, which is part of A'(n) and needs to be subject to further reduction.

Previously, the library handled this carry in the following way:
- If it was positive, the reduction routine would simply return, accepting that the output is only 'almost' reduced (down to n+1 limbs).
- If it was negative, the reduction would use the helper routine `ecp_fix_negative()` to turn compute a positive modular representative.

It is noteworthy that the accumulation of `A(i)`s always happens at 32-bit granularity, even on a 64-bit system. This is because the underlying moduli can be written in the form `p(2^32)` for an integer polynomial p, but not in the form p(2^64). For example, the prime underlying secp224r1 is `2^224 - 2^96 + 1`.

The present PR applies two simplifications:

- Firstly, it simplifies the series of addition/subtraction steps of `A(i) +- A(n+j) +- ...` in the calculation of `A'(i)`, by operating in `int64_t` throughout. On a 32-bit system, this is likely going to lead to similar assembly than our previous code, but the C-code is more readable because it does handle the carry explicitly. On a 64-bit system, accumulating in 64-bit is a native operation and thus should be faster than the previous code.

- Secondly, allow repetitions of the modular reduction sequence: After the first 'round' our representative has the form `A'(0),..,A'(n-1)`, _plus_ a signed carry out in `A'(n)`. The idea is to re-run the modular reduction sequence, but taking into account that all A'(n+j) are 0 except for j==0, in which case it's the previous round's signed carry-out.

  It can be observed that two such additional simplified rounds are enough to quasi-reduce a modulus to n limbs.

  This also removes the need for `ecp_fix_negative()`, which is removed alongside its dedicated tests in `test_suite_ecp`.

As a nice side-effect, this change removes the need for negative MPIs in the secp{192,224,256} modular reduction routines.

### CODE REVIEW

**Comment by hanno-arm** [comment]
I rebased this on top of development and made some simplifications. More documentation to come, but feel free to have a look @tom-cosgrove-arm

File: library/ecp_curves.c [hunk 1]
```diff
@@ -4931,6 +4931,7 @@ static int ecp_mod_p192( mbedtls_mpi *N )
 #define MAX32       N->n
 #define A( j )      N->p[j]
 #define STORE32     N->p[i] = cur;
+#define STORE0      N->p[i] = 0;
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/ecp_curves.c:4935]: No, you're right -- this was a preparation for a later commit which however I hadn't pushed. I flattened the PR into one change now, which seemed preferable (previous commits made changes to 32-bit code which was later made obsolete anyway through the introduction of accumulation in 64-bit).

**Comment by hanno-arm** [comment]
@mpg @tom-cosgrove-arm Could you have a look at this?

**Comment by daverodgman** [comment]
Closing; this is covered by on-going optimised reduction activities

**Pull request closed without merging**